### PR TITLE
Grid gap + individual column padding options

### PR DESCRIPTION
**Pull request by sobo-odoo**
[IMP] web_editor: clean the snippet UI before cloning it

When cloning a snippet, if it had some previews displayed, the previews
may be cloned with it. This is not a good behavior because the clone
is a new snippet so it should not have all the preview classes that were
added when editing. Moreover, these previews could also cause bugs.

To solve this, an idea could be to call the `cleanForSave` function of
all its options before cloning the snippet. But the problem is that in
some options, the `cleanForSave` function does more than simply clean
the UI (as it was originally supposed to) and so it cannot be used for
that purpose only.

As a solution, this commit adds a new function called `cleanUI` whose
exclusive purpose will be to clean the snippet UI. In the future, it
will replace the `cleanForSave` function.

task-3369695

---
[IMP] web_editor: add support for data-min/max to InputUserValueWidget

Before this commit, when using a `we-input`, there was no way to easily
specify the min and max values that could be entered in the input, which
could be useful when too low/high values could break some behaviors, for
example.

This commit adds the support for the `data-min` and `data-max` data-
attributes in `InputUserValueWidget`. More precisely, when these
attributes are added on a `we-input`, if the user writes a value smaller
than the `data-min` or bigger than the `data-max`, the value snaps to
these bounds. When using the arrows, the value cannot be decreased/
increased beyond them either.

task-3369695

---
[REF] web_editor: refactor some grid code

In the grid mode code, the code managing the transformation of a grid
item into a normal column (when it is dropped or when we toggle the
normal mode) is repeated at several places.

This commits fixes that by creating a function doing this conversion
instead.

task-3369695

---
[IMP] web_editor, website: add individual Padding option for grid items

Before this commit, the "Padding" grid mode option, added in commit [1],
was impacting all the grid items at the same time, which was not
convenient since all these elements would have the same padding and
could not be styled individually.

This commit improves this option in order to modify the padding of the
grid items individually. The highlight effect, added in commit [2], when
changing the padding is still added.

[1]: https://github.com/odoo/odoo/commit/cc406afcea7bf5846233a9f97a4a8ac5f618f3ec
[2]: https://github.com/odoo/odoo/commit/f3f2c4c73ce08d3898b8c88e1ef045f7bbd1f624

task-3369695

---
[REF] web_editor, website: remove duplicated grid/column options

When adding the grid options in commits [1] and [6], some of them have
been duplicated in order to cover the different use cases: grid-only,
columns-only and grid + columns. However, they should have been declared
only once and then called when needed.

This commit extracts the common options into views, to avoid duplicating
them.

[1]: https://github.com/odoo/odoo/commit/cc406afcea7bf5846233a9f97a4a8ac5f618f3ec
[6]: https://github.com/odoo/odoo/commit/85b352af319edec84407f2046cf795b4e5503460

task-3369695

---
[IMP] web_editor, website: add a Spacing option to modify the grid gaps

The `display: grid` property allows to modify the spacing (or gaps)
between the rows and columns of the grid with the `row-gap` and `column-
gap` CSS properties. When the grid mode was added with [1], these
properties were deliberately ignored and the gaps were forced to 0px.

This commit adds the "Spacing (Y, X)" option that allows to modify these
grid gaps. A grid preview is also added when using this option, in order
to visualize what is being changed in the grid.

[1]: https://github.com/odoo/odoo/commit/cc406afcea7bf5846233a9f97a4a8ac5f618f3ec

task-3369695

---
[FIX] web_editor: make background grid more visible on dark backgrounds

When redesigning the grid layout and handles, commit [3] removed the
filter that allowed the grid to be clearly visible on dark backgrounds
(added in commit [4]). Without this, the background grid when dragging/
resizing and the highlight effect when changing the gaps are difficult
to see.

This commit adds this filter back, but with a lowered opacity in order
to not obstruct the view too much.

[3]: https://github.com/odoo/odoo/commit/70f723a9f78c406746f5409f6a89cbe93ad21580
[4]: https://github.com/odoo/odoo/commit/1aa81243b7c80069c576381e2f3b04281455794f

task-3369695

---
[FIX] website: make the "Text Cover" snippet look good in grid mode

Before this commit, toggling the "Text Cover" snippet to grid mode did
not produce a good-looking result: the column containing the background
image loses all its padding and therefore appears smaller and shifted
compared to the normal mode.

This commit adds a background color to the image column. Indeed, with a
background color, and more precisely thanks to the `o_cc` class, the
padding is taken into account when computing the grid items size, making
them look similar to how they were in normal mode.

task-3369695

---
[FIX] website: fix the new design of the Banner snippet

In commit [5], the Banner snippet has been redesigned in order to be
in grid mode by default, to highlight some cool features of the editor.
However, some grid classes and attributes are missing or not well set:
- the first column has a `g-height-4` class despite having 8 rows: this
causes the vertical resize to not work correctly.
- the row does not have the `data-row-count` attribute, which is needed
to display the background grid correctly.

This commit fixes this by replacing the class by `g-height-8` and adding
the `data-row-count` attribute, which is set to 11 rows.

[5]: https://github.com/odoo/odoo/commit/3cbdf754ff8fac8a77887c4307b658cb86b0be1d

task-3369695

### CODE REVIEW

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 1]
```diff
@@ -5318,6 +5291,9 @@ registry.layout_column = SnippetOptionWidget.extend({
         // Removing the grid class
         rowEl.classList.remove('o_grid_mode');
         const columnEls = rowEl.children;
+        // Removing the grid previews (if any).
+        this._clearAllPreviews(columnEls);
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:5341]: Why is this necessary? Should not your `cleanUI` method be called at the right moment to handle this automatically?

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 2]
```diff
@@ -5326,20 +5302,93 @@ registry.layout_column = SnippetOptionWidget.extend({
...
+     * Removes all the eventual grid previews, to avoid unwanted behaviors (when
+     * toggling normal mode for example).
+     *
+     * @private
+     */
+    _clearAllPreviews(columnEls) {
+        // Removing the padding previews.
+        for (const columnEl of columnEls) {
+            if (columnEl._removePaddingPreview) {
+                columnEl._removePaddingPreview();
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:5369]: I don't want something like that to be attached to the column elements. I guess the right way depends on the answer to my previous comment ("can't you rely on cleanUI?")... :grimacing:

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 3]
```diff
@@ -5326,20 +5302,93 @@ registry.layout_column = SnippetOptionWidget.extend({
...
+    },
+    /**
+     * @override
+     */
+    cleanUI() {
+        // Remove the padding highlights.
+        if (this.$target[0].classList.contains("o_we_padding_highlight")) {
+            this._removePaddingPreview();
+        }
+    },
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:5408]: Maybe we should call `cleanUI` everywhere `cleanForSave`  is already called? (maybe just calling cleanUI` in the `SnippetEditor` `cleanForSave` would be enough?)

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 4]
```diff
@@ -5326,20 +5302,93 @@ registry.layout_column = SnippetOptionWidget.extend({
...
+    /**
+     * @override
+     */
+    cleanUI() {
+        // Remove the padding highlights.
+        if (this.$target[0].classList.contains("o_we_padding_highlight")) {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:5405]: I'd remove this condition

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 5]
```diff
@@ -5326,20 +5302,93 @@ registry.layout_column = SnippetOptionWidget.extend({
...
+    //--------------------------------------------------------------------------
+    // Options
+    //--------------------------------------------------------------------------
+
+    /**
+     * @override
+     */
+    async selectStyle(previewMode, widgetValue, params) {
+        await this._super(...arguments);
+        if (params.cssProperty.startsWith("--grid-item-padding")) {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:5419]: I'd prefer you'd list them both instead of relying on `startsWith`. Would be easier not to make a mistake while grepping.

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 6]
```diff
@@ -5326,20 +5302,93 @@ registry.layout_column = SnippetOptionWidget.extend({
...
+            // Reset the animation.
+            this._removePaddingPreview();
+            void this.$target[0].offsetWidth; // Trigger a DOM reflow.
+
+            // Highlight the padding when changing it, by adding a pseudo-
+            // element with an animated colored border inside the grid item.
+            this.$target[0].classList.add("o_we_padding_highlight");
+            this.$target[0]._removePaddingPreview = this._removePaddingPreview.bind(this);
+            this.$target[0].addEventListener("animationend", this.$target[0]._removePaddingPreview);
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:5428]: Worst case scenarion, your options can communication using the `notify` method

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 7]
```diff
@@ -5044,6 +5044,24 @@ registry.Box = SnippetOptionWidget.extend({
...
+    },
+    /**
+     * @override
+     */
+    cleanUI() {
+        // Remove the grid preview.
+        if (this.gridPreviewEl) {
+            this._removeGridPreview();
+        }
+    },
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js]: Same: would remove the `if` + duplicated code (cleanUI should be called at the same time as cleanForSave anyway).

File: addons/website/views/snippets/snippets.xml [hunk 8]
```diff
@@ -539,6 +539,10 @@
             <we-button class="o_we_bg_brand_primary o_grid" data-add-element="text" title="Text" data-name="text">Text</we-button>
             <we-button class="o_we_bg_brand_primary o_grid" data-add-element="button" title="Button" data-name="button">Button</we-button>
         </we-button-group>
+        <we-row string="Spacing (Y, X)">
+            <we-input data-dependencies="grid_mode" data-select-style="" data-css-property="row-gap" data-unit="px" data-prevent-important="true" data-apply-to=".row"/>
+            <we-input data-dependencies="grid_mode" data-select-style="" data-css-property="column-gap" data-unit="px" data-prevent-important="true" data-apply-to=".row"/>
+        </we-row>
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml]: Maybe time to not duplicated that code? You could extract the common layout part into a view no? :)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@sobo-odoo I added multiple comments. The first one for the first commit, I handled it myself (before deciding this should go back to some dev for you :wink:), so don't forget to fetch (I will push-force just after sending this review) :+1:

**Comment by qsm-odoo** [comment]
@sobo-odoo Push-forced, the diff I made myself: https://github.com/odoo/odoo/compare/dd590a41400b08802667a276e3ad3cfc206ad29d..57afe124f55fa97dd8198d2294db309c80954d86 (check if it's right, I may have made a mistake :shrug:). Ping me once the rest is ready  :+1:

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 1 above)
Comment (sobo-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:5341]: It is for the eventual case the user would toggle the normal mode with some previews still displayed, so the grid preview will not be considered as a child and the padding highlight classes will not be on the columns (really a corner case I think but there are tracebacks if he does that so better be safe...)  :grimacing: 

You are right, I could just call `cleanUI` there then :slightly_smiling_face: I added this function because I didn't know how to call the one of the GridColumn option but as you mentioned in an another comment, I can make them communicate with `notify` indeed :open_mouth: 
So yes the next comment will surely be fixed by that :slightly_smiling_face:  (I didn't know how to do that in another way which is why storing in the DOM was the solution I used but I knew it would not be accepted haha)

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 3 above)
Comment (sobo-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:5409]: Yes, I didn't do it because you said to just do it for the clone first, so I thought it could stay like this for the time being and that it would be done later in another PR (where the duplicate would be removed) :open_mouth: 

I will do that then :slightly_smiling_face:

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 5 above)
Comment (sobo-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:5420]: Indeed, forgot to change it when changing the other :grimacing:

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 3 above)
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:5409]: Yes, what I meant was that you did not need to check all the cleanForSave definition to extract the relevant parts into cleanUI for now. But definitely call cleanUI at the same time as cleanForSave :+1:

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 1 above)
Comment (sobo-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:5341]: In the end, I trigger a `clean_ui_request`, because it seems that the `notify` only works for options of the same editor and not from an editor to another (so here I can't communicate between a "section" option and a "column" option) 
Or did I miss something ?

**Comment by sobo-odoo** [comment]
@qsm-odoo I made the changes :slightly_smiling_face:

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 9]
```diff
@@ -5054,6 +5054,13 @@ registry.Box = SnippetOptionWidget.extend({
 
 
 registry.layout_column = SnippetOptionWidget.extend({
+    /**
+     * @override
+     */
+    cleanUI() {
+        // Remove the grid preview.
+        this._removeGridPreview();
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:5062]: I think the comment is useless :wink:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Small last comments :)

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 10]
```diff
@@ -5200,6 +5207,34 @@ registry.layout_column = SnippetOptionWidget.extend({
...
     },
+    /**
+     * @override
+     */
+    async selectStyle(previewMode, widgetValue, params) {
+        const rowEl = this.$target[0];
+        if (params.cssProperty === "column-gap") {
+            // Limit the column-gap so the row does not overflow: the gap cannot
+            // be bigger than the grid column size.
+            widgetValue = `${Math.min(parseFloat(widgetValue), rowEl.clientWidth / 23)}px`;
```
[the remark this answers is not in the document]
Comment (sobo-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:5218]: > You could round the value

Yes, I was not sure it would be justified so I let the real value and waited for an opinion about it :slightly_smiling_face: 

> but really I think we can just hardcode a maximum value, like 60px and it will be ok, no?

I thought of using a hardcoded value but the issue is I didn't know which value to use because the columns size is relative to the screen width (unlike the rows that are hardcoded at 50px) So there is always a point where the column gap could be too big :confused: 
Here at worst, when the screen will be small enough such that the columns size will be 0, there is a "guarantee" that the total width would still not overflow (this is what I had in mind when setting the ratio but I agree it is overkill :sweat_smile:) 
The other reason I used the column size as a limit is that if a user needs to put a gap bigger than the column size, then he could just move/resize the grid item by one row instead of changing the gap :shrug: 

But now that I think of it, I could use the row width when it is "container small" and divide it by 11 (gaps) (it seems to be a good upper bound) so it gives 690 / 11 = 62.73 px => rounded down it is 62px
But then 60px is a prettier number haha so yes in the end I agree with you => I will put 60px :slightly_smiling_face:  

> make that a InputUserValueWidget parameter

I wanted to add a `data-max` attribute to the input indeed, but as it was not a hardcoded value, I didn't know how to make it "dynamic", which is why I did it in `selectStyle` :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (sobo-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:5218]: Ah... we can't add `data-max` to inputs ? :open_mouth: It seems to only apply to ranges :grimacing: Should I add the support for that ?

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 11]
```diff
@@ -5336,20 +5317,69 @@ registry.layout_column = SnippetOptionWidget.extend({
...
+    async selectStyle(previewMode, widgetValue, params) {
+        await this._super(...arguments);
+        if (["--grid-item-padding-y", "--grid-item-padding-x"].includes(params.cssProperty)) {
+            // Reset the animation.
+            this._removePaddingPreview();
+            void this.$target[0].offsetWidth; // Trigger a DOM reflow.
+
+            // Highlight the padding when changing it, by adding a pseudo-
+            // element with an animated colored border inside the grid item.
+            this.$target[0].classList.add("o_we_padding_highlight");
```
[the remark this answers is not in the document]
Comment (sobo-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:5402]: Indeed, I thought I tested all the cases but it seems that undoing the gap/padding changes while the preview is still visible (must be quick) and then redoing makes the previews reappear :grimacing: (they are well removed afterwards but they stay if we save before it disappears...)  => I will add the observers :slightly_smiling_face:

About the behavior in stable, yes I noticed while doing this PR that I didn't test well all these cases and so the history does not work well (and cloning with the preview on makes it impossible to save the editor...) :sweat_smile:  
I already planned to fix this in stable later :slightly_smiling_face:

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:5218]: Yep that's what I meant by "maybe a good idea would be to make that a InputUserValueWidget parameter", I saw this was not yet a thing, would be a good thing to add :+1:

**Comment by sobo-odoo** [comment]
@qsm-odoo I made the changes :slightly_smiling_face: (see [diff](https://github.com/odoo/odoo/compare/58937f0d5c1d3377add56efe0931d38e01c0af88..961b07b1061006360e33e000e94833710781edd7))

I added a new commit about the `data-max` (and `data-min` too, while I was at it) on inputs (the second one) :slightly_smiling_face:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Final comments for info then I push-force myself and merge :)

**Review by qsm-odoo — APPROVED** [review]
As previously advertised, I had to make those changes to fix the min/max commit: https://github.com/odoo/odoo/compare/6b056114dd4528d4ce312992bb6fb31ba95281d1..b43f7e93bf0528a0b18315faa773d69f5ee2de1e
But overall: good job :clap: 

@robodoo rebase-ff r+

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 12]
```diff
@@ -1257,6 +1257,30 @@ const InputUserValueWidget = UnitUserValueWidget.extend({
...
     //--------------------------------------------------------------------------
 
+    /**
+     * @override
+     */
+    getActiveValue() {
+        const params = this._methodsParams;
+        if (!params.max && !params.min) {
+            return this._super(...arguments);
+        }
```
[the remark this answers is not in the document]
Comment (sobo-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: Ah yes indeed, I totally forgot about 0 :grimacing: Because when using the inputs, it often puts 0 when writing nothing or negative number so I thought that specifying `data-min="0"` would never be done haha 
But it is not the case for all the inputs so I should have considered it, thanks :+1:

File: addons/web_editor/static/src/scss/wysiwyg_snippets.scss [hunk 13]
```diff
@@ -2613,6 +2613,7 @@ we-select.o_grid we-toggler {
         stroke: $o-we-bg-darkest;
         stroke-opacity: .2;
         stroke-width: 1px;
+        filter: drop-shadow(-1px -1px 0px rgba(255, 255, 255, 0.3));
```
[the remark this answers is not in the document]
Comment (sobo-odoo) [reply addons/web_editor/static/src/scss/wysiwyg_snippets.scss:2616]: With this filter, there is more contrast when the background is dark, making the grid easier to see. Even if the rectangles are not that "more visible", the trick of this filter is that it shifts the shadow left and up by 1px, making the border look like a double line (a dark one and a white one) so it will always be visible depending on the background :slightly_smiling_face:  

With the filter: 
image

Without the filter:
image
